core/controllers/profile_test.py

```diff
@@ -791,6 +795,120 @@ def test_delete_delete_account_page_disabled(self):
+                    u'edited_exploration_ids': [],
+                    u'created_exploration_ids': []
+                },
+                u'general_feedback_thread_user_data': {},
+                u'question_snapshot_metadata_data': {},
+                u'general_feedback_message_data': {},
+                u'story_progress_data': {},
+                u'learner_playlist_data': {},
+                u'collection_rights_data': {
+                    u'voiced_collection_ids': [],
+                    u'owned_collection_ids': [],
+                    u'viewable_collection_ids': [],
+                    u'editable_collection_ids': []
+                },
+                u'skill_snapshot_metadata_data': {},
+                u'exploration_user_data_data': {},
+                u'collection_snapshot_metadata_data': {},
+                u'exploration_rights_data': {
+                    u'viewable_exploration_ids': [],
+                    u'owned_exploration_ids': [],
+                    u'voiced_exploration_ids': [],
+                    u'editable_exploration_ids': []
+                },
+                u'topic_snapshot_metadata_data': {},
+                u'completed_activities_data': {},
+                u'general_feedback_thread_data': {},
+                u'topic_rights_snapshot_metadata_data': {},
+                u'user_stats_data': {},
+                u'exploration_rights_snapshot_metadata_data': {},
+                u'user_subscriptions_data': {
+                    u'creator_ids': [],
+                    u'collection_ids': [],
+                    u'activity_ids': [],
+                    u'general_feedback_thread_ids': [],
+                    u'last_checked': None
+                },
+                u'config_property_snapshot_metadata_data': {},
+                u'exploration_snapshot_metadata_data': {},
+                u'incomplete_activities_data': {},
+                u'user_skill_mastery_data': {},
+                u'exp_user_last_playthrough_data': {},
+                u'user_settings_data': {
+                    u'username': u'editor',
+                    u'last_agreed_to_terms':
+                        python_utils.UNICODE(
+                            user_settings.last_agreed_to_terms),
+                    u'last_started_state_translation_tutorial': None,
+                    u'last_started_state_editor_tutorial': None,
+                    u'normalized_username': u'editor',
+                    u'first_contribution_msec': None,
+                    u'preferred_language_codes': [
+                        u'en'
+                    ],
+                    u'creator_dashboard_display_pref': u'card',
+                    u'subject_interests': [],
+                    u'default_dashboard': None,
+                    u'preferred_site_language_code': None,
+                    u'user_bio': u'',
+                    u'profile_picture_data_url':
+                        python_utils.UNICODE(
```

Comment:
All items in the expected value should be constants. user_settings.profile_picture_data_url etc. are not constants.

Do a swap on the relevant functions if you need to fix some of the values.

Reply:
Since I have to swap multiple values, does this mean I have to do the following?

```
with self.swap():
    with self.swap():
        with self.swap():
            with self.swap():
                with self.swap():
```

This seems a little weird to me

Reply:
See e.g. scripts.install_third_party_test and others for examples of how to handle the case of multiple swaps.

Reply:
@seanlip I tried swapping values for the user_settings object; however, this doesn't work because swap will swap out the object properties, whereas export_data will perform a query from the database. So the only way (that I can see) for performing this swap is changing the database values of the object prior to calling export_data. Is this the route I should take?

Reply:
Or can I just swap the data? I tried doing the model method I mentioned above but this caused a 302 Temporarily Moved error for some reason...

Reply:
You could e.g. swap out the method that generates a profile picture with a dummy implementation that uses a custom profile picture that you are defining in tests. For the datetime stuff you might be able to swap out the method that gets the current time (utcnow or something in utils.py) with a custom impl that returns a specific time that you can then assert against.

Swapping the response data or swapping the model doesn't make sense because that's part of what you're actually trying to test. The idea is to do the minimum change in the pipeline to make the result deterministic.

Reply:
Resolved